crates/evm/coverage/src/analysis.rs

```
@@ -511,127 +466,36 @@ impl SourceAnalyzer {
-                            contract_id.source_id
-                        )
-                    }),
-                    contract_id.contract_name.clone(),
-                )
-                .visit(
-                    self.contracts
-                        .get(contract_id)
-                        .unwrap_or_else(|| {
-                            panic!("We should have the AST of contract: {contract_id:?}")
-                        })
-                        .clone(),
-                )?;
-
-                let is_test = items.iter().any(|item| {
-                    if let CoverageItemKind::Function { name } = &item.kind {
-                        name.is_test()
-                    } else {
-                        false
-                    }
-                });
-
-                // Record this contract's base contracts
-                // We don't do this for test contracts because we don't care about them
-                if !is_test {
-                    self.contract_bases.insert(
-                        contract_id.clone(),
-                        base_contract_node_ids
-                            .iter()
-                            .filter_map(|base_contract_node_id| {
-                                self.contract_ids.get(base_contract_node_id).cloned()
-                            })
-                            .collect(),
-                    );
-                }
-
-                // For tests and contracts with no items we still record an empty Vec so we don't
-                // end up here again
-                if items.is_empty() || is_test {
-                    self.contract_items.insert(contract_id.clone(), Vec::new());
+            let ContractVisitor { items, .. } = ContractVisitor::new(
+                contract_id.source_id,
+                self.sources.get(&contract_id.source_id).unwrap_or_else(|| {
+                    panic!("We should have the source code for source ID {}", contract_id.source_id)
```

can we return an error?
the function returns an error